### PR TITLE
Endpoint to get gamification board data

### PR DESCRIPTION
**Pull request by Devesh21700Kumar**
**Description**

Add endpoint to retrieve gamification data
Data for gamification boards will be retrieved
Suitable endpoint and documentation has been added
Added index and ordering for the request of gamification data

Used the below link as reference for indexing collection based on a feild
https://docs.mongodb.com/manual/tutorial/sort-results-with-indexes/

image

image

**Type of Change:**

- [x] Code
- [x] New Feature
- [x] Documentation

**How Has This Been Tested?**

Terminal commands to test code:-
1). npm run lint
2). npm run lint:fix
3). nest start
4). npm run test

**Checklist**

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] My changes generate no new warnings
- [x] The title of my pull request is a short description of the requested changes.

### CODE REVIEW

File: src/modules/user/docUtils/user.responsedoc.ts [hunk 1]
```diff
@@ -123,3 +127,29 @@ export class EnrolledCourseResponseBody {
...
+
+  /**
+   * Score of the user
+   * @example 100
+   */
+  score: number;
+
+  /**
+   * The photo url
+   * @example 'https://g.gle/mypic.jpeg'
```
Comment (Abhishek-kumar09) [inline src/modules/user/docUtils/user.responsedoc.ts:152]: Add a link to dummy image that works, people copy doc fields without thinking anything so the links mentioned should work.
https://cdn.pixabay.com/photo/2015/03/04/22/35/head-659652_1280.png

File: src/modules/user/user.controller.ts [hunk 2]
```diff
@@ -121,12 +121,20 @@ export class UserController {
...
 
   @Get('/users')
+  @Roles(Role.ADMIN)
   @ApiOperation({ summary: 'Retreive user list' })
   @ApiOkResponse(responsedoc.getAllUser)
   async getAllUser() {
     return await this.userService.getAllUser();
   }
 
+  @Get('/Gamification')
```
Comment (Abhishek-kumar09) [inline src/modules/user/user.controller.ts:131]: use small letter

**Review by Abhishek-kumar09 — CHANGES_REQUESTED** [review]

File: src/modules/user/user.controller.ts (hunk 2 above)
Comment (Devesh21700Kumar) [reply src/modules/user/user.controller.ts]: @Abhishek-kumar09  fixed

File: src/middleware/preAuth.middleware.ts [hunk 3]
```diff
@@ -26,12 +26,19 @@ export class PreauthMiddleware implements NestMiddleware {
...
+              },
+            );
           } else {
             const newUser = new this.userModel({
               email,
               fid: uid,
               role: Role.STUDENT,
               photoUrl: picture,
+              log_in_time: ' ',
             });
```
Comment (Abhishek-kumar09) [inline src/middleware/preAuth.middleware.ts:42]: Extract these to function so that code is more readable

File: src/modules/user/user.service.ts [hunk 4]
```diff
@@ -37,6 +37,20 @@ export class UserService {
...
+        .find({}, { _id: false })
+        .select('first_name last_name score photoUrl')
+        .sort({ score: 1 })
+        .exec();
+      return users;
+    } catch (e) {
+      throw new InternalServerErrorException(e);
+    }
+  }
+
```
Comment (Abhishek-kumar09) [inline src/modules/user/user.service.ts:53]: The frontend is using pagination and doesn't show all the info at the same time, please add pagination here

**Review by Devesh21700Kumar — COMMENTED** [review]
image


@Abhishek-kumar09  pagination added as well for the retrieval of gamification board data

File: src/modules/user/user.service.ts (hunk 4 above)
Comment (Devesh21700Kumar) [reply src/modules/user/user.service.ts]: @Abhishek-kumar09 Pagination Added

File: src/app.module.ts [hunk 5]
```diff
@@ -51,6 +51,10 @@ export class AppModule {
           path: 'course/(.*)',
           method: RequestMethod.GET,
         },
+        {
+          path: 'user/Gamification',
```
Comment (Abhishek-kumar09) [inline src/app.module.ts:55]: small letters

File: src/modules/user/docUtils/user.responsedoc.ts [hunk 6]
```diff
@@ -123,3 +133,29 @@ export class EnrolledCourseResponseBody {
...
+
+  /**
+   * Score of the user
+   * @example 100
+   */
+  score: number;
+
+  /**
+   * The photo url
+   * @example 'https://cdn.pixabay.com/photo/2015/03/04/22/35/head-659652_1280.png'
```
Comment (Abhishek-kumar09) [inline src/modules/user/docUtils/user.responsedoc.ts:158]: create bit.ly link for this

File: src/modules/user/schema/user.schema.ts [hunk 7]
```diff
@@ -44,9 +44,15 @@ export class User {
...
   fId: string;
+
+  @Prop()
+  log_in_time: string;
 }
 
-export const UserSchema = SchemaFactory.createForClass(User);
+const UserSchema = SchemaFactory.createForClass(User);
+
+UserSchema.index({ score: 1 });
```
Comment (Abhishek-kumar09) [inline src/modules/user/schema/user.schema.ts:54]: descending order

File: src/modules/user/user.service.ts [hunk 8]
```diff
@@ -37,6 +37,21 @@ export class UserService {
...
   }
 
+  // fetch all gamification data
+  async getAllGamified(skipNum: string): Promise<User[]> {
+    try {
+      const skip = parseInt(skipNum, 10);
+      const users = await this.userModel
+        .find({}, { _id: false }, { skip: skip, limit: 10 })
+        .select('first_name last_name score photoUrl')
+        .sort({ score: 1 })
```
Comment (Abhishek-kumar09) [inline src/modules/user/user.service.ts:47]: descending order

**Pull request merged**